Feature/provide requester secret refs in deal params

jbern0rd opened the pull request:
Ground work needed to handle requester secret references through `TaskDescription` in `iexec-sms`

jbern0rd committed:
Provide requester secrets references in DealParams, retrieve them in TaskDescription

jbern0rd committed:
Use slf4j in RequestOrder, replace printStackTrace with logging event

jbern0rd committed:
Add RequestOrderTest class

mcornaton commented on src/main/java/com/iexec/common/task/TaskDescription.java:
```diff
@@ -142,6 +144,11 @@ public static TaskDescription toTaskDescription(String chainTaskId,
             inputFiles = chainDeal.getParams().getIexecInputFiles();
         }
+        Map<String, String> secrets = Map.of();
+        if (chainDeal.getParams() != null &&
```
Testing whether `chainDeal.getParams()` is null there is pretty strange as we use it a bunch of times without any caution later - [like here](https://github.com/iExecBlockchainComputing/iexec-common/blob/6273701ec7bc749f55d6f7cbe5b3c873f27d2aed/src/main/java/com/iexec/common/task/TaskDescription.java#L159).
So we're sure it'll throw an NPE if `chainDeal.getParams()` is null. What about adding a single test at the beginning of the method so that we're sure this value is never null later? BTW, that's already what's done for `chainDeal`.

jbern0rd replied:
It would be possible if `chainDeal.getParams()` was the only call which could throw a NPE.
NPE can be thrown by the following calls: `getBotSize()`, `getChainApp()`, `getChainCategory()`, `getChainDataset()`
It is not possible to do a single test at the beginning with those multiple calls.
**I will remove the unnecessary check and add a TODO.**

Later, I would like to investigate how data read from the chain are populated in Object instances.
It would be great to guarantee that `chainDeal.getParams()` can never be null.

What does everyone think of this suggestion ?

zguesmi replied:
Actually `chainDeal.getParams()` cannot be null. See how it is created in the deal [here](https://github.com/iExecBlockchainComputing/iexec-common/blob/63f66080c6ab269897a7d3bd125479b3b6b0eef1/src/main/java/com/iexec/common/chain/IexecHubAbstractService.java#L403). Cf [here](https://github.com/iExecBlockchainComputing/iexec-common/blob/63f66080c6ab269897a7d3bd125479b3b6b0eef1/src/main/java/com/iexec/common/chain/ChainDeal.java#L71).

zguesmi replied:
Also just for information, the deal params field onchain is not guaranteed to be non-null. So the object `DealParams` will be defined but the content of [iexecArgs](https://github.com/iExecBlockchainComputing/iexec-common/blob/63f66080c6ab269897a7d3bd125479b3b6b0eef1/src/main/java/com/iexec/common/chain/ChainDeal.java#L81) could potentially be null which is not necessarily a problem.

jbern0rd replied:
I will take another look to improve my understanding of the code.
Thanks for the help.

jbern0rd replied:
See https://github.com/iExecBlockchainComputing/iexec-common/pull/298/commits/f3b818be8003c042ebf381f70553e7290a43b315

mcornaton commented on src/test/java/com/iexec/common/sdk/order/payload/RequestOrderTest.java:
```diff
@@ -0,0 +1,67 @@
+
+    @Test
+    void shouldSerializeParametersWithEncryptedResult() {
+        for (String storageProxy : List.of(DealParams.DROPBOX_RESULT_STORAGE_PROVIDER, DealParams.IPFS_RESULT_STORAGE_PROVIDER)) {
```
What about a parametrized test there, instead of a `for` loop?

jbern0rd replied:
Fixed in https://github.com/iExecBlockchainComputing/iexec-common/pull/298/commits/7107a8db754fd4807bf1ec149afd78fa41506bd1

mcornaton commented on src/test/java/com/iexec/common/sdk/order/payload/RequestOrderTest.java:
```diff
@@ -0,0 +1,67 @@
+    void shouldSerializeSecrets() {
+        String expectedData = "{\"iexec_secrets\":{\"0\":\"0\"}}";
+        DealParams params = DealParams.builder()
+                .iexecSecrets(Map.of("0", "0"))
```
I'd prefer to have two distinct values. Otherwise, there could be some switch between key and value and we could not notice that.

james-toussaint replied:
Yes, plus a map having more than 1 mapping

jbern0rd replied:
Fixed in https://github.com/iExecBlockchainComputing/iexec-common/pull/298/commits/7107a8db754fd4807bf1ec149afd78fa41506bd1

james-toussaint commented on src/main/java/com/iexec/common/chain/DealParams.java:
```diff
@@ -50,6 +51,10 @@
     @JsonProperty("iexec_result_storage_proxy")
     private String iexecResultStorageProxy;
 
+    // Keys are app secrets indices, values are requester secrets indices
```
```suggestion
    /*
    The "iexec_secrets" key -> value map declares which secrets of the requester 
    can be accessed by a targeted TEE application.
    - A value should hold the secret name of a secret belonging to the requester.
    This secret must be previously pushed on the SMS. 
    - A key a associated to a value indicates how the TEE application can access 
    that value.
    
    Example:
    
    - A requester has a secret value "2000-01-01" pushed on the SMS. The name of 
    this secret is "my-birthday".
    - A requester has a secret value "+331234567890" pushed on the SMS. The name 
    of this secret is "my-phone-number".

    The application developer creates a TEE application which allows requesters 
    to :
    1. Share their birthday by affecting their secret to position "2"
    2. Share their phone number by affecting their secret to position "5"

    In order to share birthday and phone number to the TEE application, the 
    requester should declare:
    "iexec_secrets" : { "2": "my-birthday" , "5": "my-phone-number" }
    
    By doing this, the TEE application code will be able to read following 
    environment variables:
    IEXEC_REQUESTER_SECRET_2=2000-01-01
    IEXEC_REQUESTER_SECRET_5=+331234567890
    */
```

jbern0rd replied:
I agree with your suggestions, plus:
* I will try to add it as a javadoc comment
* Maybe the last part should be documented in `iexec-sms,` `iexec-common` has no knowledge of `IEXEC_REQUESTER_SECRET_` prefix

james-toussaint replied:
Yes, I agree about responsibility of the last part. We can either :
- Add a note stating env var prefix is fake and just for overall understandability purposes
- Relocate that part to `com.iexec.common.utils.IexecEnvUtils`

zguesmi replied:
I think it's a good idea to add `IEXEC_REQUESTER_SECRET_` to `IexecEnvUtils`.

jbern0rd replied:
I have a question about this.
Do I just add the String definition to `IexecEnvUtils` ?
`public static final String IEXEC_REQUESTER_SECRET_PREFIX = "IEXEC_REQUESTER_SECRET_";`
Do I add `IEXEC_APP_DEVELOPER_SECRET_PREFIX` as well ?

zguesmi replied:
We can add both yes. Even though the variables will probably be used only in the SMS, it's a good idea to have them all at the same place.

jbern0rd replied:
See 23794c6  and 7040309

james-toussaint replied:
Thanks!

zguesmi commented on src/test/java/com/iexec/common/sdk/order/payload/RequestOrderTest.java:
```diff
@@ -0,0 +1,67 @@
+import java.util.List;
+import java.util.Map;
+
+class RequestOrderTest {
```
Can we add a test that serializes all params at once?

jbern0rd replied:
Fixed in https://github.com/iExecBlockchainComputing/iexec-common/pull/298/commits/7107a8db754fd4807bf1ec149afd78fa41506bd1

jbern0rd committed:
DealParams instance is never null when executing toTaskDescription

IexecHubAbstractService#getChainDeal is called before.
In this method, ChainDeal#getStartTime cannot be null, which means
ChainDeal#stringToDealParams was called and has returned a DealParams instance.

jbern0rd committed:
Fix and add test methods in RequestOrderTest

james-toussaint commented on src/main/java/com/iexec/common/chain/ChainDeal.java:
```diff
@@ -72,15 +73,19 @@ public static DealParams stringToDealParams(String params) {
             return DealParams.builder()
                     .iexecArgs(params)
-                    .iexecInputFiles(new ArrayList<>())
+                    .iexecInputFiles(List.of())
```
```suggestion
                    .iexecInputFiles(Collections.emptyList())
```
?
https://stackoverflow.com/a/39400310

jbern0rd replied:
Fixed in https://github.com/iExecBlockchainComputing/iexec-common/pull/298/commits/c642ed8631c33d38f36a6e45da21850c8272bf56

james-toussaint commented on src/main/java/com/iexec/common/chain/ChainDeal.java:
```diff
@@ -72,15 +73,19 @@ public static DealParams stringToDealParams(String params) {
-                    .iexecInputFiles(new ArrayList<>())
+                    .iexecInputFiles(List.of())
                     .iexecDeveloperLoggerEnabled(false)
+                    .iexecSecrets(Map.of())
```
```suggestion
                    .iexecSecrets(Collections.emptyMap())
```

jbern0rd replied:
Fixed in https://github.com/iExecBlockchainComputing/iexec-common/pull/298/commits/c642ed8631c33d38f36a6e45da21850c8272bf56

james-toussaint commented on build.gradle:
```diff
@@ -61,9 +61,7 @@ dependencies {
-    testImplementation 'org.junit.jupiter:junit-jupiter-api:5.6.0'
-    testImplementation 'org.junit.jupiter:junit-jupiter-engine:5.6.0'
-    testImplementation 'org.mockito:mockito-core:3.6.0'
+    testImplementation 'org.junit.jupiter:junit-jupiter:5.6.0'
```
:+1:

zguesmi commented on src/test/java/com/iexec/common/chain/ChainDealTest.java:
```diff
@@ -62,26 +63,57 @@
+        Assertions.assertNull(deal.getStartTime());
+        Assertions.assertNull(deal.getTag());
+        Assertions.assertNull(deal.getTrust());
+        Assertions.assertNull(deal.getWorkerStake());
```
You could use `hasAllNullFieldsOrProperties()` from `org.assertj.core.api.Assertions.assertThat`.

jbern0rd replied:
Fixed in https://github.com/iExecBlockchainComputing/iexec-common/pull/298/commits/86b25b4e02c469cd822bf7e25a2dfd36e10c0e5b

zguesmi commented on src/test/java/com/iexec/common/chain/ChainDealTest.java:
```diff
@@ -62,26 +63,57 @@
         DealParams params = stringToDealParams(ARGS);
         assertEquals(ARGS, params.getIexecArgs());
         assertEquals(0, params.getIexecInputFiles().size());
+        assertEquals(0,params.getIexecSecrets().size());
```
```suggestion
        assertEquals(0, params.getIexecSecrets().size());
```

jbern0rd replied:
Fixed in https://github.com/iExecBlockchainComputing/iexec-common/pull/298/commits/86b25b4e02c469cd822bf7e25a2dfd36e10c0e5b

zguesmi commented on src/test/java/com/iexec/common/sdk/order/payload/RequestOrderTest.java:
```diff
@@ -48,18 +63,47 @@ void shouldSerializeArgs() {
+        Map<String, String> secrets = Map.of("0", "secretA", "2", "secretD");
+        String serializedSecrets = secrets.entrySet().stream()
+                .map(entry -> "\"" + entry.getKey() + "\":\"" + entry.getValue() + "\"")
+                .collect(Collectors.joining(","));
```
I find this more readable:
```suggestion
        String serializedSecrets = "{\"0\":\"secretA\", \"2\":\"secretD\"}"
```

jbern0rd replied:
Sadly, I don't have much of a choice here.
The keys are unordered in a Map.
Serialized data can thus be `{"0":"secretA","2":"secretD"}` or `{"2":"secretD","0":"secretA"}`
That's why I decided to perform my own serialization with a stream...

zguesmi replied:
I see. For that reason I generally prefer reading `expectedData` in a `JSONObject` and compare its fields.
No problem, we can keep it that way.

mcornaton replied:
What if you provide a `LinkedHashMap` instead? Its order should be predictable.
See https://docs.oracle.com/javase/8/docs/api/java/util/LinkedHashMap.html.

jbern0rd replied:
I think what we really care is proving something like:
`stringToDealParams(toStringParams(dealParams)).equals(dealParams)` but is not the same java Object

`ChainDeal#toStringParams` could be refactored to `DealParams#toString`
`RequestOrder#toStringParams` could be refactored to `DealParams#parseString`

To provide the maximum value, I would prefer to do this kind of refactoring instead of trying to check the correct serialization and deserialization of hard-coded json strings.

jbern0rd replied:
See 23794c6

jbern0rd committed:
Use Collections.emptyList and Collections.emptyMap in ChainDeal

Those methods create respectively an immutable List and Map.
For each method, the returned instance is created only once.
It avoids to create a new List or Map for each call.

jbern0rd committed:
Fix ChainDealTest

zguesmi reviewed (approved):
Well done!

mcornaton reviewed (approved):
LGTM! :)

jbern0rd committed:
Refactor ChainDeal#stringToDealParams and RequestOrder#toStringParams

jbern0rd committed:
Add secret prefix variables for requester and app developer

zguesmi reviewed (approved):
Clean!

james-toussaint commented on src/main/java/com/iexec/common/chain/DealParams.java:
```diff
@@ -58,5 +121,48 @@
+     *                    In the second case, every other parameters will have default values.
+     * @return the created instance
+     */
+    public static DealParams createFromString(String paramString) {
```
:+1:

james-toussaint commented on src/test/java/com/iexec/common/chain/DealParamsTest.java:
```diff
@@ -0,0 +1,189 @@
+
+    @ParameterizedTest
+    @ValueSource(strings = {DealParams.DROPBOX_RESULT_STORAGE_PROVIDER, DealParams.IPFS_RESULT_STORAGE_PROVIDER})
+    void testSerializationWithAllParams(String storageProvider) {
```
:+1:

james-toussaint reviewed (approved):
Thanks :)

mcornaton commented on src/main/java/com/iexec/common/chain/DealParams.java:
```diff
@@ -17,11 +17,69 @@
+ * DealParams allow a requester to provide <b>args</b>, <b>input files</b>, <b>secrets</b>,
+ * <b>result storage configuration</b>... Those configurations are written on the blockchain
+ * when a deal is submitted.
+ * <p>
```
I usually prefer to use `<br>` tag. Don't know what @jeremyjams and @zguesmi prefer.

james-toussaint replied:
b stands for bold :)

jbern0rd replied:
I followed this documentation https://www.oracle.com/technical-resources/articles/java/javadoc-tool.html
I like `<p>` which stands for paragraph but `<br>` is fine as well.

james-toussaint replied:
OK comment was on `<p>` ..!

Default Intellij likes `<p>` too:
![image](https://user-images.githubusercontent.com/33313130/149810600-d9c1154e-3891-4f60-a8ba-a4eb31fc26bc.png)

mcornaton replied:
If both of you prefer `<p>`, let's use that tag so we're consistent :)

mcornaton reviewed (approved):
Looks great, thanks!

jbern0rd committed:
Merge branch 'develop' into feature/provide-requester-secret-refs-in-deal-params

jbern0rd committed:
Move deprecated jcenter repository to last position
